### PR TITLE
Vecsim - integrate ad-hoc BF optimization

### PR DESCRIPTION
**Pull request by alonre24**
This PR introduces an important optimisation for hybrid vector similarity query. That is, if evaluates which of the two modes will lead to better runtime of the query: 
1, Using a batch iterator that returns the next best results from the index upon demand.
2. Computing distances for every vector who represents a document that passes the filter only, and take the best result. 
To determine this, the index is using a heuristics based on decision tree classifiers that is available in the updated version of the `Vector Similarity` dependency to which RediSearch now points to.

Changes were made in some of the hybrid query tests to cover both modes (+ use `ft.debug` to verify the mode in which the last query ran in).

This PR also introduces - 
1. python notebook with a demo for hybrid-queries
2. the code for the benchmark that was used to train the classifiers (todo: decide if this is the appropriate place)

### CODE REVIEW

File: src/hybrid_reader.c [hunk 1]
```diff
@@ -117,23 +122,52 @@ static void alternatingIterate(HybridIterator *hr, VecSimQueryResult_Iterator *v
...
+void computeDistances(HybridIterator *hr) {
+  float upper_bound = INFINITY;
+  RSIndexResult *cur_res = hr->base.current;
+  RSIndexResult *cur_child_res;  // This will use the memory of hr->child->current.
+  RSIndexResult *cur_vec_res = NewDistanceResult();
+
+  while (hr->child->Read(hr->child->ctx, &cur_child_res) != INDEXREAD_EOF) {
+    float dist = (float)VecSimIndex_GetDistanceFrom(hr->index, cur_child_res->docId, hr->query.vector);
+    // If this id is not in the vector index (since it was deleted), dist will return as NaN.
+    if (isnanf(dist)) {
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:134]: do we have such cases? can you refer to a test?

File: src/hybrid_reader.c [hunk 2]
```diff
@@ -117,23 +122,52 @@ static void alternatingIterate(HybridIterator *hr, VecSimQueryResult_Iterator *v
...
   VecSimBatchIterator *batch_it = VecSimBatchIterator_New(hr->index, hr->query.vector);
   float upper_bound = INFINITY;
   while (VecSimBatchIterator_HasNext(batch_it)) {
-    size_t batch_size = hr->query.k;  // todo: add heuristics here
+    hr->numIterations++;
+    size_t vec_index_size = VecSimIndex_IndexSize(hr->index);
+    size_t n_res_left = hr->query.k - heap_count(hr->topResults);
+    // Since NumEstimated(child) is an upper bound, it can be higher than index size, then we increase
+    // batch and make sure that it is at least one.
+    size_t batch_size = n_res_left * ((float)vec_index_size / hr->child->NumEstimated(hr->child->ctx)) + 1;
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:170]: please check that `hr->child->NumEstimated(hr->child->ctx)` is not zero

File: src/hybrid_reader.c [hunk 3]
```diff
@@ -287,16 +317,22 @@ IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KN
...
   if (child_it == NULL) {
-    hi->mode = STANDARD_KNN;
-//  } else if (UseBF(child_it->NumEstimated(child_it), query, index)) {
-//    hi->mode = HYBRID_ADHOC_BF;
+    hi->mode = VECSIM_STANDARD_KNN;
   } else {
-    hi->mode = HYBRID_BATCHES;
-    //Todo: apply heuristics (batch_size = k / (vec_index_size*child_it->NumEstimated(child_it)))
+    size_t subset_size = child_it->NumEstimated(child_it->ctx);
+    if (subset_size > VecSimIndex_IndexSize(index)) {
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:326]: when could it be?

File: src/hybrid_reader.c [hunk 4]
```diff
@@ -248,6 +277,7 @@ static void HR_Rewind(void *ctx) {
       IndexResult_Free(hr->returnedResults[i]);
     }
     array_clear(hr->returnedResults);
+    hr->child->Rewind(hr->child->ctx);
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:280]: why the rewind addition?

File: tests/cpptests/test_cpp_index.cpp [hunk 5]
```diff
@@ -699,15 +701,39 @@ TEST_F(IndexTest, testHybridVector) {
   ASSERT_EQ(VecSimIndex_IndexSize(index), max_id);
 
   float query[] = {(float)max_id, (float)max_id, (float)max_id, (float)max_id};
-  KNNVectorQuery knn_query = {.vector = query, .vecLen = d, .k = 10, .order = BY_ID};
+  KNNVectorQuery top_k_query = {.vector = query, .vecLen = d, .k = 10, .order = BY_SCORE};
```
Comment (DvirDukhan) [inline tests/cpptests/test_cpp_index.cpp:704]: knn_query seems fine, no?

File: src/hybrid_reader.c (hunk 1 above)
Comment (alonre24) [reply src/hybrid_reader.c:134]: yes, `test_hybrid_query_non_vector_score` for example. when we update some id, the doc id remains in the index (until garbage collector is removing it) but not in the vector index

File: src/hybrid_reader.c (hunk 3 above)
Comment (alonre24) [reply src/hybrid_reader.c:326]: It happens in union for example. This isn't accurate, it is an upper bound.

File: tests/cpptests/test_cpp_index.cpp (hunk 5 above)
Comment (alonre24) [reply tests/cpptests/test_cpp_index.cpp:704]: yep I dont know why i changed it...

File: tests/pytests/test_vecsim.py [hunk 6]
```diff
@@ -605,15 +689,15 @@ def test_hybrid_query_batches_non_vector_score(env):
     # All documents should match, so TOP 10 takes the 10 with the largest ids. Since we sort by default score
     # and "value" is optional, expect that 100 will come first, and the rest will be sorted by id in ascending order.
     expected_res_1 = [10L, '100', '3', ['__v_score', '0', 't', 'other'], '91', '2', ['__v_score', '10368', 't', 'text value'], '92', '2', ['__v_score', '8192', 't', 'text value'], '93', '2', ['__v_score', '6272', 't', 'text value'], '94', '2', ['__v_score', '4608', 't', 'text value'], '95', '2', ['__v_score', '3200', 't', 'text value'], '96', '2', ['__v_score', '2048', 't', 'text value'], '97', '2', ['__v_score', '1152', 't', 'text value'], '98', '2', ['__v_score', '512', 't', 'text value'], '99', '2', ['__v_score', '128', 't', 'text value']]
-    execute_hybrid_query(env, '((text ~value)|other)=>[KNN 10 @v $vec_param]', query_data, 't', sort_by_vector=False).equal(expected_res_1)
+    execute_hybrid_query(env, '((text ~value)|other)=>[KNN 10 @v $vec_param]', query_data, 't', sort_by_vector=False, batches_mode=False).equal(expected_res_1)
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:692]: how come batch mode is false here? what triggered BF?

File: docs/Vector-Similarity-Hybrid-Queries.ipynb [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+      "      Successfully uninstalled redis-4.2.0rc1\n",
+      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
+      "rltest 0.5.0 requires redis<5.0.0,>=4.1.2, but you have redis 3.5.3 which is incompatible.\u001b[0m\u001b[31m\n",
+      "\u001b[0mSuccessfully installed redis-3.5.3\n",
+      "Requirement already satisfied: numpy in /home/alon/.local/lib/python3.8/site-packages (1.21.4)\n"
+     ]
+    }
+   ],
+   "source": [
+    "!pip install git+https://github.com/RediSearch/redisearch-py.git@params\n",
```
Comment (DvirDukhan) [inline docs/Vector-Similarity-Hybrid-Queries.ipynb:70]: you can use redis-py now
we will submit this PR now as-is and modify all ipynb files with new redis-py API after

File: tests/pytests/test_vecsim.py [hunk 8]
```diff
@@ -393,159 +393,241 @@ def test_memory_info(env):
...
+    for i in range(10):
+        expected_res.append(str(index_size-i))
+        expected_res.append(['__v_score', str(dim*i**2), 't', 'other' if i % 5 == 0 else 'text value'])
+    execute_hybrid_query(env, '(@t:other|text)=>[KNN 10 @v $vec_param]', query_data, 't').equal(expected_res)
+
+    # Expect for top 10 results from vector search that still has the original text "text value".
+    expected_res = [10L]
+    res_count = 0
+    for i in range(index_size):
+        # The desired ids are the top 10 ids that do not divide by 5.
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/pytests/test_vecsim.py:476]: `range(13)` works

File: tests/pytests/test_vecsim.py [hunk 9]
```diff
@@ -393,159 +393,241 @@ def test_memory_info(env):
...
 
-    expected_res_3 = [5L, '49', ['__v_score', '332928', 'num', '49'], '48', ['__v_score', '346112', 'num', '48'], '47', ['__v_score', '359552', 'num', '47'], '46', ['__v_score', '373248', 'num', '46'], '45', ['__v_score', '387200', 'num', '45']]
-    execute_hybrid_query(env, '(@num:[45 (50])=>[KNN 10 @v $vec_param]', query_data, 'num').equal(expected_res_3)
+    # Expect to get results with maximum numeric value of index_size-100
+    expected_res = [10L]
+    for i in range(10):
+        expected_res.append(str(index_size-100-i))
+        expected_res.append(['__v_score', str(dim*(100+i)**2), 'num', str(index_size-100-i)])
+    expected_res_2 = [10L, '1000', ['__v_score', '100000000', 'num', '1000'], '999', ['__v_score', '100040000', 'num', '999'], '998', ['__v_score', '100080016', 'num', '998'], '997', ['__v_score', '100120032', 'num', '997'], '996', ['__v_score', '100160064', 'num', '996'], '995', ['__v_score', '100200096', 'num', '995'], '994', ['__v_score', '100240144', 'num', '994'], '993', ['__v_score', '100280192', 'num', '993'], '992', ['__v_score', '100320256', 'num', '992'], '991', ['__v_score', '100360320', 'num', '991']]
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/pytests/test_vecsim.py:589]: yes, thanks

File: tests/pytests/test_vecsim.py (hunk 6 above)
Comment (alonre24) [reply tests/pytests/test_vecsim.py:692]: Because the index size is smaller than in other tests (100)

File: tests/pytests/test_vecsim.py [hunk 10]
```diff
@@ -185,7 +185,7 @@ def testCreate(env):
     for _ in env.retry_with_rdb_reload():
         info = [['identifier', 'v', 'attribute', 'v', 'type', 'VECTOR']]
         assertInfoField(env, 'idx1', 'attributes', info)
-        env.assertEqual(env.cmd("FT.DEBUG", "VECSIM_INFO", "idx1", "v")[:-1], ['ALGORITHM', 'HNSW', 'TYPE', 'FLOAT32', 'DIMENSION', 1024L, 'METRIC', 'IP', 'INDEX_SIZE', 0L, 'M', 16L, 'EF_CONSTRUCTION', 200L, 'EF_RUNTIME', 10L, 'MAX_LEVEL', -1L, 'ENTRYPOINT', -1L, 'MEMORY'])
+        env.assertEqual(env.cmd("FT.DEBUG", "VECSIM_INFO", "idx1", "v")[:-3], ['ALGORITHM', 'HNSW', 'TYPE', 'FLOAT32', 'DIMENSION', 1024L, 'METRIC', 'IP', 'INDEX_SIZE', 0L, 'M', 16L, 'EF_CONSTRUCTION', 200L, 'EF_RUNTIME', 10L, 'MAX_LEVEL', -1L, 'ENTRYPOINT', -1L, 'MEMORY'])
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/pytests/test_vecsim.py]: Sorry, I didn't push it before. Now it is changed.

**Review by DvirDukhan — APPROVED** [review]

File: src/hybrid_reader.c (hunk 1 above)
Comment (ashtul) [reply src/hybrid_reader.c:134]: Why change to float?

File: src/hybrid_reader.c [hunk 11]
```diff
@@ -117,23 +122,55 @@ static void alternatingIterate(HybridIterator *hr, VecSimQueryResult_Iterator *v
...
+  }
+  IndexResult_Free(cur_vec_res);
+}
+
+static void prepareResults(HybridIterator *hr) {
+    if (hr->mode == VECSIM_STANDARD_KNN) {
       hr->list =
           VecSimIndex_TopKQuery(hr->index, hr->query.vector, hr->query.k, &(hr->runtimeParams), hr->query.order);
       hr->iter = VecSimQueryResult_List_GetIterator(hr->list);
     return;
```
Comment (ashtul) [inline src/hybrid_reader.c:153]: space

File: src/hybrid_reader.c (hunk 3 above)
Comment (ashtul) [reply src/hybrid_reader.c:329]: Please leave a comment with the exlanation

File: src/hybrid_reader.c [hunk 12]
```diff
@@ -287,16 +320,22 @@ IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KN
   hi->base.isValid = 1;
   hi->list = NULL;
   hi->iter = NULL;
+  hi->numIterations = 0;
```
Comment (ashtul) [inline src/hybrid_reader.c:323]: Let's add `numIterations` to FT.PROFILE

File: src/hybrid_reader.c [hunk 13]
```diff
@@ -228,6 +259,7 @@ static size_t HR_LastDocId(void *ctx) {
 static void HR_Rewind(void *ctx) {
   HybridIterator *hr = ctx;
   hr->resultsPrepared = false;
+  hr->numIterations = 0;
```
Comment (ashtul) [inline src/hybrid_reader.c:262]: I think it is better to have here `numIterations++` so developers can see it in FT.PROFILE.

File: src/hybrid_reader.c [hunk 14]
```diff
@@ -117,23 +122,55 @@ static void alternatingIterate(HybridIterator *hr, VecSimQueryResult_Iterator *v
...
   }
   // Batches mode.
+  if (hr->child->NumEstimated(hr->child->ctx) == 0) {
+    return;
+  }
   VecSimBatchIterator *batch_it = VecSimBatchIterator_New(hr->index, hr->query.vector);
   float upper_bound = INFINITY;
   while (VecSimBatchIterator_HasNext(batch_it)) {
-    size_t batch_size = hr->query.k;  // todo: add heuristics here
+    hr->numIterations++;
```
Comment (ashtul) [inline src/hybrid_reader.c:168]: Where is it used?

File: src/hybrid_reader.c [hunk 15]
```diff
@@ -287,16 +320,22 @@ IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KN
...
+      hi->mode = VECSIM_HYBRID_ADHOC_BF;
+    } else {
+      hi->mode = VECSIM_HYBRID_BATCHES;
+    }
   }
-  if (hi->mode != STANDARD_KNN) {
+  if (hi->mode != VECSIM_STANDARD_KNN) {
     hi->topResults = rm_malloc(heap_sizeof(query.k));
     heap_init(hi->topResults, cmpVecSimResByScore, NULL, query.k);
     hi->returnedResults = array_new(RSIndexResult *, query.k);
```
Comment (ashtul) [inline src/hybrid_reader.c:341]: Can be joined with the previous block

File: src/hybrid_reader.c [hunk 16]
```diff
@@ -239,7 +271,7 @@ static void HR_Rewind(void *ctx) {
   hr->lastDocId = 0;
   hr->base.isValid = 1;
 
-  if (hr->mode != STANDARD_KNN) {
+  if (hr->mode != VECSIM_STANDARD_KNN) {
```
Comment (ashtul) [inline src/hybrid_reader.c:274]: use `==` for readibility

**Review by ashtul — COMMENTED** [review]
Looks good
Some comments

File: docs/Vector-Similarity-Hybrid-Queries.ipynb [hunk 17]
```diff
@@ -0,0 +1,363 @@
+{
```
[the remark this answers is not in the document]
Comment (alonre24) [reply docs/Vector-Similarity-Hybrid-Queries.ipynb:1]: It is the hybrid query demo, we want to save it...

File: src/hybrid_reader.c (hunk 13 above)
Comment (alonre24) [reply src/hybrid_reader.c:262]: Agreed. Will change

File: src/hybrid_reader.c (hunk 14 above)
Comment (alonre24) [reply src/hybrid_reader.c:168]: For monitoring only. Will move to `ft.profile`

File: tests/cpptests/test_cpp_index.cpp [hunk 18]
```diff
@@ -220,11 +223,11 @@ TEST_P(IndexFlagsTest, testRWFlags) {
 
 INSTANTIATE_TEST_SUITE_P(IndexFlagsP, IndexFlagsTest, ::testing::Range(1, 32));
 
-InvertedIndex *createIndex(int size, int idStep) {
+InvertedIndex *createIndex(int size, int idStep, int start_with=0) {
   InvertedIndex *idx = NewInvertedIndex((IndexFlags)(INDEX_DEFAULT_FLAGS), 1);
 
   IndexEncoder enc = InvertedIndex_GetEncoder(idx->flags);
-  t_docId id = idStep;
+  t_docId id = start_with>0 ? start_with : idStep;
```
Comment (DvirDukhan) [inline tests/cpptests/test_cpp_index.cpp:230]: space

File: src/hybrid_reader.c (hunk 14 above)
Comment (ashtul) [reply src/hybrid_reader.c:168]: What does `numIterations` means here?
IMO it is better to increase every rewind. Maybe add additional `numRewind`?

File: tests/pytests/test_profile.py [hunk 19]
```diff
@@ -231,13 +231,13 @@ def testProfileVector(env):
   env.assertEqual(env.cmd("FT.DEBUG", "VECSIM_INFO", "idx", "v")[-1], 'HYBRID_ADHOC_BF')
 
   for i in range(6, 10001):
-    conn.execute_command('hset', str(i), 'v', 'babababa', 't', "hello world")
+    conn.execute_command('hset', str(i), 'v', 'bababada', 't', "hello world")
```
Comment (DvirDukhan) [inline tests/pytests/test_profile.py:234]: why the change of the query? scenario for 2 batches?

Comment (alonre24) [reply tests/pytests/test_profile.py:234]: Yes... To make these further from the query vector than `1`, but closer than `5`

File: src/hybrid_reader.c (hunk 14 above)
Comment (alonre24) [reply src/hybrid_reader.c:168]: In the `fr.profile` I named it "Batches number" - it is the number of batches that were given from the vector index (which is also the number of rewinds that we did to the child iterator. I think that "Batches number" is a bit more informative than "numRewind". I will add that to the docs

**Review by DvirDukhan — APPROVED** [review]

**Review by ashtul — DISMISSED** [review]
Great job

**Review by DvirDukhan — APPROVED** [review]
